docshare.go

```diff
@@ -188,23 +189,43 @@ func fetchPermission(ctx context.Context, client *http.Client, docID, email stri
 	return fetchPermission(ctx, client, docID, email, pageToken(p.NextPageToken))
 }
 
-func errorResponse(res *http.Response) error {
-	var e struct {
-		Error struct {
-			Errors []struct{ Message string }
+func doRetry(ctx context.Context, client *http.Client, req *http.Request) (*http.Response, error) {
+	var nretry int
+	for {
+		res, err := client.Do(req.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+		if res.StatusCode >= 200 && res.StatusCode <= 299 {
+			return res, nil
+		}
+
+		err = errorResponse(res)
+		res.Body.Close()
+		if !isRetriable(res.StatusCode, err) {
+			return nil, err
+		}
+		nretry++
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(backoff(nretry)):
+			// Retry.
 		}
 	}
-	b, err := ioutil.ReadAll(res.Body)
-	if err != nil {
-		return errors.New(res.Status)
+}
+
+func backoff(n int) time.Duration {
+	const max = 10 * time.Second
```

Comment:
Just a somewhat educated guess based on the API quotas, request latency and experience with `claat export`.
Let's say Drive API call takes 1s to return, on average. The API has 10 QPS (queries per second) per user quota by default. Users can request more quota though.

You have `maxConcurrecy` global constant set to 10. So, this should be around as slow as it can get given the Drive API constraints.

Could also make it `time.Duration(maxConcurrency) * time.Second` instead of `10 * time.Second`, if you think it makes more sense. If one day you set `maxConcurrency` to a 100 though, this would be too much. I would stick with 10s - it's 10x the API call latency and by that time we're already deep at the 5th retry. It's more likely to never succeed at that point anyway.

We're doing the same backoff in `claat export` but without any artificial concurrency limit and it has always worked so far.